README.md
# pycololight

A Python3 wrapper for interacting with LifeSmart Cololight.

Supports the following cololight devices:

- hexagon
- strip

## Usage

```python
from pycololight import PyCololight

# Setup hexagon device
light = PyCololight(device="hexagon", host="1.1.1.1")

# Setup strip device, and include dynamic effects
light = PyCololight(device="strip", host="1.1.1.1", dynamic_effects=True)

# Turn on at 60% brightness
light.on = 60

# Set brightness to 70%
light.brightness = 70

# Set light colour
light.colour = (255, 127, 255)

# Set effect
light.effect = "Sunrise"

# Create custom effect
light.add_custom_effect(
  name="custom effect",
  colour_scheme="Shadow",
  colour="Red, Yellow",
  cycle_speed=11,
  mode=1
)

# Turn off
light.on = 0
```

Mapping of modes for custom effects can be found [here](https://github.com/BazaJayGee66/pycololight/blob/main/MODES.md)


CHANGELOG.md
# 2.1.0

## New Features

- Add ability to set timeout for device

# 2.0.0

This release enables support for the cololight strip device.

## New Features

- Add support for the cololight strip device
- Add dynamic effects for the cololight strip device
- Add custom effects for the cololight strip device

## Breaking Changes

- A device (`hexagon` or `strip`) must now be passed when creating `PyCololight` object

# 1.0.0

## New Features

- Add state to get light state (on/off) and brightness - Thanks [@pim12345](https://github.com/pim12345)

## Fixes

- Add missing exceptions
- Add BrightnessException if brightness not between 0 and 100
- Refactor for simpler code
- Rename `include_default_effects` to `restore_default_effects`
- Remove `count` property
- Add type hints, and docstring

# 0.0.2

## New Features

## Fixes

- Refactor for simpler code

# 0.0.1

Initial Release

## New Features

- Turn On
- Turn Off
- Set brightness
- Set colour
- Set effect
- Include/Exclude default effects
- Add custom effect

## Fixes


tests/test_effects.py
import pytest

from pycololight import (
    Effects,
    ColourSchemeException,
    ColourException,
    CycleSpeedException,
    ModeExecption,
)  # pylint: disable=import-error

from pycololight.constants import (
    STRIP_DYANMIC_EFFECTS,
)  # pylint: disable=import-error


class TestEffects:
    def test_default_effects_returns_dict_default_effects(self):
        effects = Effects(device="hexagon")

        expected_default_effects = {
            "80s Club": ["049a0000"],
            "Cherry Blossom": ["04940800"],
            "Cocktail Parade": ["05bd0690"],
            "Instagrammer": ["03bc0190"],
            "Pensieve": ["04c40600"],
            "Savasana": ["04970400"],
            "Sunrise": ["01c10a00"],
            "The Circus": ["04810130"],
            "Unicorns": ["049a0e00"],
            "Christmas": ["068b0900"],
            "Rainbow Flow": ["03810690"],
            "Music Mode": ["07bd0990"],
        }

        assert effects.default_effects == expected_default_effects

    def test_custom_effect_colour_schemes_returns_supported_colour_schemes(self):
        effects = Effects(device="hexagon")

        supported_colour_schemes = [
            "Breath",
            "Shadow",
            "Flash",
            "Flicker",
            "Scene",
            "Mood",
            "Selected",
        ]

        assert effects.custom_effect_colour_schemes() == supported_colour_schemes

    def test_custom_effect_colour_scheme_colours_returns_colour_scheme_colours(self):
        effects = Effects(device="hexagon")

        expected_colours = [
            "Red, Green, Blue",
            "Rainbow",
            "Green",
            "Azure",
            "Blue",
            "Purple",
            "Red",
            "Orange",
            "Yellow",
            "White",
        ]

        assert (
            effects.custom_effect_colour_scheme_colours("Flicker") == expected_colours
        )

    def test_colour_hex_raises_exception_when_bad_scheme(self):
        effects = Effects(device="hexagon")

        with pytest.raises(ColourSchemeException):
            effects._colour_hex("bad_scheme", "colour", 1)

    def test_colour_hex_raises_exception_when_bad_colour(self):
        effects = Effects(device="hexagon")

        with pytest.raises(ColourException):
            effects._colour_hex("Mood", "bad_colour", 1)

    def test_cycle_speed_hex_raises_exception_when_bad_speed(self):
        effects = Effects(device="hexagon")

        with pytest.raises(CycleSpeedException):
            effects._cycle_speed_hex(35, 1)

    def test_mode_hex_raises_exception_when_bad_mode(self):
        effects = Effects(device="hexagon")

        with pytest.raises(ModeExecption):
            effects._mode_hex(0)

    def test_cycle_speed_hex_returns_hex_value(self):
        effects = Effects(device="hexagon")

        expected_responses = ["01", "20", "0d", "0b"]

        test_cycle_speeds = [(32, 1), (1, 1), (20, 1), (2, 2)]

        for index, test_speed in enumerate(test_cycle_speeds):
            cycle_speed = test_speed[0]
            mode = test_speed[1]
            cycle_speed_hex = effects._cycle_speed_hex(cycle_speed, mode)
            assert cycle_speed_hex == expected_responses[index]

    def test_mode_hex_returns_tuple_of_hex_values(self):
        effects = Effects(device="hexagon")

        expected_responses = [("05", "10"), ("05", "80"), ("06", "10"), ("06", "70")]

        test_modes = [3, 8, 17, 26]

        for index, mode in enumerate(test_modes):
            mode_hex = effects._mode_hex(mode)
            assert mode_hex == expected_responses[index]

    def test_colour_hex_returns_hex_value_for_hexagon(self):
        effects = Effects(device="hexagon")

        expected_responses = ["80", "06", "a6", "b4", "bf", "c1", "c3"]

        test_colours = [
            ["Breath", "Red, Green, Blue", 1],
            ["Breath", "Red", 13],
            ["Flicker", "Azure", 1],
            ["Mood", "Orange", 1],
            ["Selected", "Savasana", 1],
            ["Selected", "Sunrise", 1],
            ["Selected", "Unicorns", 1],
        ]

        for index, colour in enumerate(test_colours):
            colour_hex = effects._colour_hex(colour[0], colour[1], colour[2])
            assert colour_hex == expected_responses[index]

    def test_colour_hex_returns_hex_value_for_strip(self):
        effects = Effects(device="strip")

        expected_responses = ["80", "a6", "34", "c1", "c3"]

        test_colours = [
            ["Breath", "Red, Green, Blue", 1],
            ["Flicker", "Azure", 1],
            ["Mood", "Orange", 10],
            ["Selected", "Sunrise", 13],
            ["Selected", "Unicorns", 1],
        ]

        for index, colour in enumerate(test_colours):
            colour_hex = effects._colour_hex(colour[0], colour[1], colour[2])
            assert colour_hex == expected_responses[index]

    def test_custom_effect_command_returns_effect_command_hex(self):
        effects = Effects(device="hexagon")

        effect_colour_schema = "Mood"
        effect_colour = "Orange"
        effect_cycle_speed = 11
        effect_mood = 1

        custom_effect_command = effects.custom_effect_command(
            effect_colour_schema,
            effect_colour,
            effect_cycle_speed,
            effect_mood,
        )

        assert custom_effect_command == "01b41600"

    def test_custom_effect_command_returns_effect_hex_when_mode_is_2(self):
        effects = Effects(device="hexagon")

        effect_colour_schema = "Mood"
        effect_colour = "Orange"
        effect_cycle_speed = 3
        effect_mood = 2

        custom_effect_command = effects.custom_effect_command(
            effect_colour_schema,
            effect_colour,
            effect_cycle_speed,
            effect_mood,
        )

        assert custom_effect_command == "0213b400"

    def test_dynamic_effects_returns_dict_dynamic_effects(self):
        hexagon_effects = Effects(device="hexagon")
        strip_effects = Effects(device="strip")

        expected_hexagon_dynamic_effects = {}
        expected_strip_dynamic_effects = STRIP_DYANMIC_EFFECTS

        assert hexagon_effects.dynamic_effects == expected_hexagon_dynamic_effects
        assert strip_effects.dynamic_effects == expected_strip_dynamic_effects


tests/test_pycololight.py
import pytest

from pycololight import (  # pylint: disable=import-error
    PyCololight,
    BrightnessException,
    UnsupportedDeviceException,
)

from pycololight.constants import STRIP_DYANMIC_EFFECTS

from unittest.mock import patch


class TestPyCololight:
    def test_raises_exception_when_device_not_supported(self):
        with pytest.raises(UnsupportedDeviceException):
            PyCololight(device="fake", host="1.1.1.1")

    @patch("pycololight.PyCololight._send")
    def test_turn_on(self, mock_send):
        light = PyCololight(device="hexagon", host="1.1.1.1")
        assert light.on == False

        light.on = 60

        mock_send.assert_called_with(
            [
                b"SZ00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x03\x01\xcf<"
            ]
        )

        assert light.on == True
        assert light.brightness == 60

    @patch("pycololight.PyCololight._send")
    def test_setting_brightness(self, mock_send):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        light.brightness = 60

        mock_send.assert_called_with(
            [
                b"SZ00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x03\x01\xcf<"
            ]
        )

        assert light.brightness == 60

    @patch("pycololight.PyCololight._send")
    def test_setting_brightness_raises_exception_when_outside_bound(self, mock_send):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        with pytest.raises(BrightnessException):
            light.brightness = 130

    @patch("pycololight.PyCololight._send")
    def test_setting_colour(self, mock_send):
        light = PyCololight(device="hexagon", host="1.1.1.1")
        assert light.colour == None

        light.colour = (255, 127, 255)

        mock_send.assert_called_with(
            [
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x00\xff\x7f\xff"
            ]
        )

        assert light.colour == (255, 127, 255)

    @patch("pycololight.PyCololight._send")
    def test_setting_effect(self, mock_send):
        light = PyCololight(device="hexagon", host="1.1.1.1")
        assert light.effect == None

        light.effect = "Sunrise"

        mock_send.assert_called_with(
            [
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x01\xc1\n\x00"
            ]
        )
        assert light.effect == "Sunrise"

    def test_effects_returns_list_of_effects(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")
        default_effects = [
            "80s Club",
            "Cherry Blossom",
            "Cocktail Parade",
            "Instagrammer",
            "Pensieve",
            "Savasana",
            "Sunrise",
            "The Circus",
            "Unicorns",
            "Christmas",
            "Rainbow Flow",
            "Music Mode",
        ]

        assert light.effects == default_effects

    @patch("pycololight.PyCololight._send")
    def test_turn_off(self, mock_send):
        light = PyCololight(device="hexagon", host="1.1.1.1")
        light._on = True

        light.on = 0
        mock_send.assert_called_with(
            [
                b"SZ00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x03\x01\xce\x1e"
            ]
        )

        assert light.on == False

    def test_add_custom_effect_adds_effect(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        effect_name = "test_effect"
        effect_colour_schema = "Mood"
        effect_colour = "Orange"
        effect_cycle_speed = 11
        effect_mood = 1

        light.add_custom_effect(
            effect_name,
            effect_colour_schema,
            effect_colour,
            effect_cycle_speed,
            effect_mood,
        )

        assert effect_name in light.effects
        assert light._effects[effect_name] == ["01b41600"]

    def test_add_custom_effect_adds_effect_when_mode_is_2(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        effect_name = "test_effect"
        effect_colour_schema = "Mood"
        effect_colour = "Orange"
        effect_cycle_speed = 3
        effect_mood = 2

        light.add_custom_effect(
            effect_name,
            effect_colour_schema,
            effect_colour,
            effect_cycle_speed,
            effect_mood,
        )

        assert effect_name in light.effects
        assert light._effects[effect_name] == ["0213b400"]

    def test_custom_effect_colour_schemes_returns_supported_colour_schemes(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        supported_colour_schemes = [
            "Breath",
            "Shadow",
            "Flash",
            "Flicker",
            "Scene",
            "Mood",
            "Selected",
        ]

        assert light.custom_effect_colour_schemes() == supported_colour_schemes

    def test_custom_effect_colour_scheme_colours_returns_colour_scheme_colours(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        expected_colours = [
            "Red, Green, Blue",
            "Rainbow",
            "Green",
            "Azure",
            "Blue",
            "Purple",
            "Red",
            "Orange",
            "Yellow",
            "White",
        ]

        assert light.custom_effect_colour_scheme_colours("Flicker") == expected_colours

    def test_excluding_defualt_effects(self):
        light = PyCololight(device="hexagon", host="1.1.1.1", default_effects=False)

        assert light.effects == []

    def test_default_effects_returns_list_of_default_effects(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        default_effects = [
            "80s Club",
            "Cherry Blossom",
            "Cocktail Parade",
            "Instagrammer",
            "Pensieve",
            "Savasana",
            "Sunrise",
            "The Circus",
            "Unicorns",
            "Christmas",
            "Rainbow Flow",
            "Music Mode",
        ]

        assert light.default_effects == default_effects

    def test_dynamic_effects_returns_list_of_dynamic_effects(self):
        hexagon_light = PyCololight(device="hexagon", host="1.1.1.1")
        strip_light = PyCololight(device="strip", host="1.1.1.1")

        hexagon_dynamic_effects = []
        strip_dynamic_effects = list(STRIP_DYANMIC_EFFECTS.keys())

        assert hexagon_light.dynamic_effects == hexagon_dynamic_effects
        assert strip_light.dynamic_effects == strip_dynamic_effects

    def test_restore_default_effects_adds_given_default_effects(self):
        light = PyCololight(device="strip", host="1.1.1.1", default_effects=False)

        effects = ["Pensieve", "Savasana", "Sunrise", "Tron"]

        light.restore_effects(effects)

        assert light.effects == effects

    def test_counter_returns_boolen_counter(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        assert light._counter == 1

        light._toggle_counter()
        assert light._counter == 2

        light._toggle_counter()
        assert light._counter == 1

    def test_get_config_returns_config_for_command(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        first_command_call = light._get_config("command")
        second_command_call = light._get_config("command")

        assert (
            first_command_call
            == "535a3030000000000020000000000000000000000000000000000100000000000000000004010301c"
        )
        assert (
            second_command_call
            == "535a3030000000000020000000000000000000000000000000000200000000000000000004010301c"
        )

    def test_get_config_returns_config_for_effect(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        first_effect_call = light._get_config("effect")
        second_effect_call = light._get_config("effect")

        assert (
            first_effect_call
            == "535a3030000000000023000000000000000000000000000000000100000000000000000004010602ff"
        )
        assert (
            second_effect_call
            == "535a3030000000000023000000000000000000000000000000000200000000000000000004010602ff"
        )

    def test_get_config_returns_config_for_state(self):
        light = PyCololight(device="hexagon", host="1.1.1.1")

        first_effect_call = light._get_config("state")
        second_effect_call = light._get_config("state")

        assert (
            first_effect_call
            == "535a303000000000001e000000000000000000000000000000000100000000000000000003020101"
        )
        assert (
            second_effect_call
            == "535a303000000000001e000000000000000000000000000000000200000000000000000003020101"
        )

    @patch("pycololight.PyCololight._send")
    @patch("pycololight.PyCololight._receive")
    def test_state_updates_state_and_brightness(self, mock_receive, mock_send):
        light = PyCololight(device="hexagon", host="1.1.1.1")
        assert light.on == False

        mock_receive.return_value = b"SZ00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x03\x01\xcf<"

        light.state

        mock_send.assert_called_with(
            [
                b"SZ00\x00\x00\x00\x00\x00\x1e\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x02\x01\x01"
            ],
            response=True,
        )

        assert light.on == True
        assert light.brightness == 60

    @patch("pycololight.PyCololight._send")
    def test_setting_dynamic_effect(self, mock_send):
        light = PyCololight(device="strip", host="1.1.1.1", dynamic_effects=True)

        assert light.effect == None

        light.effect = "Graffiti"

        mock_send.assert_called_with(
            [
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x93\xfe\xfe\xfe",
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x93\x05\x08\x8c",
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x93\xf0\xa0\x03",
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x93\xfa\x00\xff",
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x93\xff\xaf\x00",
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x93\x00\xc3\xff",
                b"SZ00\x00\x00\x00\x00\x00#\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x01\x06\x02\xff\x93\xff\x00\x91",
            ]
        )

        assert light.effect == "Graffiti"


pycololight/effects.py
from .constants import (
    CUSTOM_EFFECT_COLOURS,
    HEXAGON_CUSTOM_EFFECT_MODES,
    STRIP_CUSTOM_EFFECT_MODES,
    DEFAULT_EFFECTS,
    STRIP_DYANMIC_EFFECTS,
)


class CycleSpeedException(Exception):
    pass


class ColourSchemeException(Exception):
    pass


class ColourException(Exception):
    pass


class ModeExecption(Exception):
    pass


class Effects:
    """
    Effects for cololight devices
    """

    def __init__(self, device) -> None:
        self._device = device
        self._default_effects = DEFAULT_EFFECTS.copy()
        self._dynamic_effects = self._device_dynamic_effects()
        self._modes = self._device_modes()

    def _cycle_speed_hex(self, cycle_speed, mode):
        if not 1 <= cycle_speed <= 32:
            raise CycleSpeedException
        if mode in [2]:
            # Mode 2 only has speeds 1, 2, 3, which are mapped differently to other modes
            cycle_speed_values = [3, 11, 19]
            cycle_speed_value = cycle_speed_values[min(3, cycle_speed) - 1]
        else:
            cycle_speed_value = list(reversed(range(33)))[cycle_speed - 1]

        cycle_speed_hex = "{:02x}".format(cycle_speed_value)
        return cycle_speed_hex

    def _colour_hex(self, colour_scheme, colour, mode):
        if colour_scheme not in self.custom_effect_colour_schemes():
            raise ColourSchemeException
        if colour not in self.custom_effect_colour_scheme_colours(colour_scheme):
            raise ColourException

        starting_decimal = CUSTOM_EFFECT_COLOURS[colour_scheme]["decimal"]
        colour_key = CUSTOM_EFFECT_COLOURS[colour_scheme]["colours"].index(colour)
        if self._device_mode_lower(mode):
            # These modes have a lower starting decimal of 128
            starting_decimal = starting_decimal - 128
        colour_decimal = starting_decimal + colour_key
        colour_hex = "{:02x}".format(colour_decimal)
        return colour_hex

    def _mode_hex(self, mode):
        if not 1 <= mode <= len(self._modes):
            raise ModeExecption

        return self._modes[mode - 1]

    def _device_dynamic_effects(self):
        if self._device == "strip":
            dynamic_effects = STRIP_DYANMIC_EFFECTS.copy()
        else:
            dynamic_effects = {}

        return dynamic_effects

    def _device_modes(self):
        if self._device == "hexagon":
            modes = HEXAGON_CUSTOM_EFFECT_MODES.copy()
        if self._device == "strip":
            modes = STRIP_CUSTOM_EFFECT_MODES.copy()

        return modes

    def _device_mode_lower(self, mode):
        hexagon_lower_modes = [13, 14, 15, 22, 23, 24]
        strip_lower_modes = [10]

        if self._device == "hexagon":
            lower_modes = hexagon_lower_modes
        if self._device == "strip":
            lower_modes = strip_lower_modes

        return mode in lower_modes

    @property
    def default_effects(self) -> dict:
        """
        Returns a dict of default effects with names and command hex values for the device.
        """
        return self._default_effects

    @property
    def dynamic_effects(self) -> dict:
        """
        Returns a dict of dynamic effects with names and command hex values for the device.
        """
        return self._dynamic_effects

    @property
    def effects(self) -> dict:
        """
        Return a dict of all effects with names and command hex values for the device.
        """
        all_effects = {**self.default_effects, **self.dynamic_effects}
        return all_effects

    def custom_effect_command(
        self, colour_scheme: str, colour: str, cycle_speed: int, mode: int
    ) -> str:
        """
        Returns command hex for custom effect.
        """
        cycle_speed_hex = self._cycle_speed_hex(int(cycle_speed), int(mode))
        colour_hex = self._colour_hex(colour_scheme, colour, int(mode))
        mode_hex = self._mode_hex(int(mode))

        if mode in [2]:
            # Mode 2 has bytes arranged differently to other modes
            custom_effect_hex = (
                f"{mode_hex[0]}{cycle_speed_hex}{colour_hex}{mode_hex[1]}"
            )
        else:
            custom_effect_hex = (
                f"{mode_hex[0]}{colour_hex}{cycle_speed_hex}{mode_hex[1]}"
            )

        return custom_effect_hex

    def custom_effect_colour_schemes(self) -> list:
        """
        Returns a list of the available colour schemes for custom efects.
        """
        return list(CUSTOM_EFFECT_COLOURS.keys())

    def custom_effect_colour_scheme_colours(self, colour_scheme) -> list:
        """
        Returns a list of the available colours for a given colour schemes for custom efects.
        """
        return list(filter(None, CUSTOM_EFFECT_COLOURS[colour_scheme]["colours"]))


pycololight/constants.py
COMMAND_PREFIX = "535a30300000000000"

CUSTOM_EFFECT_COLOURS = {
    "Breath": {
        "decimal": 128,
        "colours": (
            "Red, Green, Blue",
            "Rainbow",
            "Green",
            "Azure",
            "Blue",
            "Purple",
            "Red",
            "Orange",
            "Yellow",
            "White",
            "Green, Blue",
        ),
    },
    "Shadow": {
        "decimal": 138,
        "colours": (
            "Red, Yellow",
            "Red, Green",
            "Red, Purple",
            "Red, Blue",
            "Green, Yellow",
            "Green, Azure",
            "Green, Blue",
            "Blue, Azure",
            "Blue, Purple",
            "Yellow, White",
            "Red, White",
            "Green, White",
            "Azure, White",
            "Blue, White",
            "Purple, White",
        ),
    },
    "Flash": {
        "decimal": 153,
        "colours": (
            "Red, Green, Blue",
            "Rainbow",
            "Green",
            "Azure",
            "Blue",
            "Purple",
            "Red",
            "Orange",
            "Yellow",
            "White",
        ),
    },
    "Flicker": {
        "decimal": 163,
        "colours": (
            "Red, Green, Blue",
            "Rainbow",
            "Green",
            "Azure",
            "Blue",
            "Purple",
            "Red",
            "Orange",
            "Yellow",
            "White",
        ),
    },
    "Scene": {
        "decimal": 173,
        "colours": (
            "Birthday",
            "Girlfriends",
            "Friends",
            "Workmates",
            "Family",
            "Lover",
        ),
    },
    "Mood": {
        "decimal": 179,
        "colours": (
            "Red",
            "Orange",
            "Yellow",
            "Green",
            "Grass",
            "Azure",
            "Blue",
            "Pink",
            "Gold",
            "Color",
            "True Color",
        ),
    },
    "Selected": {
        "decimal": 191,
        "colours": ("Savasana", "", "Sunrise", "", "Unicorns"),
    },
}

HEXAGON_CUSTOM_EFFECT_MODES = [
    ("01", "00"),
    ("02", "00"),
    ("05", "10"),
    ("05", "30"),
    ("05", "40"),
    ("05", "50"),
    ("05", "70"),
    ("05", "80"),
    ("05", "90"),
    ("05", "a0"),
    ("05", "b0"),
    ("05", "c0"),
    ("05", "00"),
    ("05", "20"),
    ("05", "30"),
    ("06", "00"),
    ("06", "10"),
    ("06", "20"),
    ("06", "30"),
    ("06", "50"),
    ("05", "f0"),
    ("05", "10"),
    ("05", "40"),
    ("05", "50"),
    ("06", "60"),
    ("06", "70"),
    ("06", "80"),
]

STRIP_CUSTOM_EFFECT_MODES = [
    ("01", "00"),
    ("02", "00"),
    ("05", "10"),
    ("05", "30"),
    ("05", "40"),
    ("05", "50"),
    ("05", "70"),
    ("05", "c0"),
    ("05", "f0"),
    ("05", "10"),
    ("06", "00"),
    ("06", "30"),
    ("06", "60"),
]

DEFAULT_EFFECTS = {
    "80s Club": ["049a0000"],
    "Cherry Blossom": ["04940800"],
    "Cocktail Parade": ["05bd0690"],
    "Instagrammer": ["03bc0190"],
    "Pensieve": ["04c40600"],
    "Savasana": ["04970400"],
    "Sunrise": ["01c10a00"],
    "The Circus": ["04810130"],
    "Unicorns": ["049a0e00"],
    "Christmas": ["068b0900"],
    "Rainbow Flow": ["03810690"],
    "Music Mode": ["07bd0990"],
}

STRIP_DYANMIC_EFFECTS = {
    "Graffiti": [
        "93fefefe",
        "9305088c",
        "93f0a003",
        "93fa00ff",
        "93ffaf00",
        "9300c3ff",
        "93ff0091",
    ],
    "Color train": ["069a0d00"],
    "Snow": ["058908c0"],
    "White link": ["06a20e00"],
    "Tron": [
        "93fefefe",
        "9304058b",
        "93f0b003",
        "93ff0000",
        "930087ff",
        "9300ff64",
    ],
    "Urban beauty": [
        "93fefefe",
        "930304b4",
        "93f04000",
        "93ff00c3",
        "9300ffc3",
    ],
    "Cyberpunk": [
        "93fefefe",
        "9303052c",
        "93f0a801",
        "93ff00d2",
        "9300c3ff",
    ],
    "Tropic affair": [
        "93fefefe",
        "93050713",
        "93f08003",
        "9319ff00",
        "93fff000",
        "93ff9b00",
        "93ff0000",
    ],
    "Go by": [
        "93fefefe",
        "93030448",
        "93f06000",
        "93ff0000",
        "9300faff",
    ],
    "Break dance": [
        "93fefefe",
        "9303048c",
        "93f0b003",
        "936400ff",
        "93ff8700",
    ],
    "Children park": [
        "93fefefe",
        "93050896",
        "93f0a602",
        "93ff00b4",
        "93b900ff",
        "9300aaff",
        "93ffeb00",
    ],
    "Thunder": [
        "93fefefe",
        "9303064a",
        "93f0a582",
        "930037ff",
        "93ffffff",
    ],
    "Red meteor": [
        "93fefefe",
        "93020237",
        "93f04800",
        "93ff0000",
    ],
    "Green meteor": [
        "93fefefe",
        "93020246",
        "93f04800",
        "9300ff0f",
    ],
    "Blue meteor": [
        "93fefefe",
        "9302027d",
        "93f04800",
        "930046ff",
    ],
    "Forest": [
        "93fefefe",
        "9303045a",
        "93f01580",
        "9332ff00",
        "93ffa500",
    ],
    "Cherry lake": [
        "93fefefe",
        "9303050d",
        "93f01680",
        "93ff738e",
        "9304ff84",
    ],
    "World cup": [
        "93fefefe",
        "9304057f",
        "93f02200",
        "930037ff",
        "93ffe100",
        "930089ce",
    ],
    "Sunset": [
        "93fefefe",
        "930404dd",
        "93f00500",
        "936400ff",
        "93ff0000",
        "93ff8700",
    ],
    "Tide": [
        "93fefefe",
        "93040775",
        "93f09a82",
        "930050ff",
        "933d98ff",
        "9395b2ff",
    ],
    "Lilac": [
        "93fefefe",
        "930407f8",
        "93f00800",
        "93fdb7ff",
        "93cb34ff",
        "93d977ff",
    ],
    "Witch hut": [
        "93fefefe",
        "93030381",
        "93f02500",
        "936900ff",
        "9300ff05",
    ],
    "Cranberry": [
        "93fefefe",
        "93030388",
        "93f01580",
        "93ff0000",
        "9300ff05",
    ],
    "Tricolor": [
        "93fefefe",
        "93040563",
        "93f08803",
        "93ff0000",
        "930087ff",
        "9300ff64",
    ],
    "Color star": ["05810cc0"],
    "Lover": ["05ba0150"],
    "Treasure": ["06bb0330"],
    "Heavy traffic": ["06a01d00"],
    "Weird story": ["06900e00"],
    "Green field": ["06950160"],
    "Seaman": ["06960100"],
    "Purple link": ["069e0e00"],
    "Blue link": ["069d0e00"],
    "Orange star": ["05870f70"],
    "Green star": ["05820f70"],
    "Red star": ["05860f70"],
    "Dusk": ["06920f00"],
    "Innocence": [
        "93fefefe",
        "93060c28",
        "93f01300",
        "935ab4ff",
        "93ff7878",
        "93a9ffac",
        "93d696ff",
        "93fff477",
    ],
    "Red alert": ["06a90100"],
    "Blue alert": ["06a70100"],
}
